### DISCUSSION TITLE
item.create Filter (Blocking) Flow creates two records

### DISCUSSION THREAD

**Discussion by lack-of-gravitas** [directus/directus#15524] [Archive]
### Describe the Bug

This is probably a PEBCAK issue so apologies in advance but the docs are not helping me here. Setting up a standard item.create blocking flow to automatically update a field value at the time a record is created, creates two records (instead of the expected one) when record is created.

**Expected Result**
Single record is created with "calculated" / supplied value in Flow set against the field on the record being created.

### To Reproduce

1 - create Event Hook flow with Filter (Blocking) on a single collection
![image](https://user-images.githubusercontent.com/10094964/189524732-07ef2e2a-3e08-450a-8979-ba29bc59fb86.png)

2 - set operation to be Create Data and specify payload
![image](https://user-images.githubusercontent.com/10094964/189524740-f6887fb1-379e-400b-ba0f-621d6a780c06.png)

3 - execute flow by creating a new record on collection 

**Actual Result**
2 records are created the original triggered from the UI and a new one triggered by the Flow. **See video snippet.**
![Peek 2022-09-11 21-29](https://user-images.githubusercontent.com/10094964/189524980-854a81fc-accb-4aa2-a968-1e771e734bcc.gif)

![image](https://user-images.githubusercontent.com/10094964/189524559-951b80f6-8e7a-4994-adc6-cc1715c59d2c.png)

### Errors Shown

None

### What version of Directus are you using?

9.15.1

### What version of Node.js are you using?

Node LTS 16.15.0

### What database are you using?

PostgreSQL 14.1 on Supabase

### What browser are you using?

Firefox Developer Edition 102.0b3 (64-bit)

### How are you deploying Directus?

Docker on fly.io

**Comment by kadiresen** [comment]
Your created flow item is firing AFTER your voucher item is created. So, when you click save, Voucher item is creating, your items.create flow is firing and creating a brand new voucher item by your flow. Seems there is no any bug in your flow definition. Your flow running as expected. Can you explain your scenario detailed you want to do?

**Comment by lack-of-gravitas** [comment]
@kadiresen yep agree that's what it's doing

What I'm trying to get done is to have the user set _value_ field and then on save, the Flow to set _balance_ and _used_ fields to 0.00 upon save of the record (commit). The intent is to store unique vouchers (and later update _balance_ and _used_ fields based on use of the value).

And based on this specific section in the docs I presume I have to set Event Hook Action to Filter (blocking).
![image](https://user-images.githubusercontent.com/10094964/189620848-8afbf6fe-99fe-4b57-bfc9-0e8c036d1d97.png)


I **think** I got it to work by changing the Flow Event Hook to **Action (non-blocking)** but I don't grok why -- 
![image](https://user-images.githubusercontent.com/10094964/189620160-e8d2208f-5b28-429a-a7e4-3725ad55717f.png)

![image](https://user-images.githubusercontent.com/10094964/189620356-34053749-0c61-4cf8-8d0c-f3cab73b2622.png)

**Reply by br41nslug** [reply]
Canceling a `filter` event within flows is still problematic afaik. What i think is a better approach is to use an `action` trigger for the create event and use `Update Data` to enhance the created item with your calculated field.

**Reply by jakedowns** [reply]
~~Filter (blocking) seems like a misnomer then. and the docs are misleading, if you can't in fact alter the payload before it is written / block the payload from being written entirely. To me, Filter (blocking) seems like a hook opportunity to add custom Filtering logic, like say, validation, or (in my case) the ability to implement Upsert logic using flows, checking if a record already exists, updating it, and canceling the original create operation.~~

ok so you can cancel the operation https://github.com/directus/directus/discussions/15655#discussioncomment-3744186

but there's now way to like gracefully cancel it without throwing an error /shrug

**Reply by rijkvanzanten** [reply]
> Filter (blocking) seems like a misnomer then. and the docs are misleading, if you can't in fact alter the payload before it is written / block the payload from being written entirely

Hmm that's exactly what it's there for. Filter events run _before_ the payload is saved to the database, so you can cancel out by throwing, or alter the payload before it's saved 👍🏻

**Reply by jakedowns** [reply]
> > Filter (blocking) seems like a misnomer then. and the docs are misleading, if you can't in fact alter the payload before it is written / block the payload from being written entirely
> 
> Hmm that's exactly what it's there for. Filter events run _before_ the payload is saved to the database, so you can cancel out by throwing, or alter the payload before it's saved 👍🏻

yup, i missed the line in the Docs about tailing reject conditional or throwing an exception in a run script block.
it's a little unintuitive from the end user's perspective tho, cause they just have to accept the error and leave the create page

it would be nice if there was some way to flag an OP as "blocked successfully" or "blocked intentionally" lol, but i understand that's probably a far off edge case

in my case i have the run script throw Error("updated record {{id}} instead.")

tho, when i use that Error to reject the original op. it seems to block my Update Data operation from committing too

![image](https://user-images.githubusercontent.com/1683122/199612429-b184b92a-7e36-4283-9761-5f2f1902bd3d.png)

i'm guessing i might just need to use a Action flow instead, and just delete the duplicate records for the given day 

ALSO when i block using a run script error... nothing is in the activity log for that run of the flow :(

if i try to block by transforming the original op payload to an empty object, it seems to do my Update AND the original create

and if i add a Condition at the end of the flow with bogus fields in the rules, it seems to still do the update and create :G

---

now i'm wondering if it's not possible to perform an arbitrary/ancillary DB operation while still blocking the original,
maybe i could trigger another action to do that part?
or, maybe i should just use an Action and delete the dupe items :G

---

ended up just going with the delete dupes Action for now

tho, while i was setting it up, i had a moment where the flow requested to delete 2 records by ID, and the operation deleted ALL records! o_O

![image](https://user-images.githubusercontent.com/1683122/199616556-10ba281e-bb40-46e0-b176-90de5f2e3229.png)
